cro/cro.py

```diff
@@ -131,44 +133,74 @@ def broadcastspawning(self, REEF, REEFpob):
         ESlarvae = np.concatenate([ESlarvae1, ESlarvae2])
         return ESlarvae
 
-    def brooding(self, REEF, REEFpob, type_brooding='op_mutation'):
+    
+    def _larvaemutation(self, larvae, pos, delta=1):
+        """
+        Description:
+            This function modifies the larvae with mutation 
+        Input:
+            - larvae: new individuals to be mutated
+            - pos: selected positions to be mutated
+            - delta: represents an increment or decrement in each mutation (it could be placed as arg in param_grid)
+        Output:
+            - larvae: modified individuals
+        """
+        (nlarvaes, b) = larvae.shape
+        MM = np.zeros([nlarvaes, b], int) # Mutation matrix
+        # int just for disc mode, replace when cont mode takes place
+
+        for key, value in self.param_grid.items():
+            m = value[0]
+            M = value[1]        
+
+        inc = (M - larvae[range(nlarvaes), pos])
+        dec = (larvae[range(nlarvaes), pos] -m) 
+        Inc = np.where(inc>dec)  
+        Dec = np.where(inc<=dec) 
+        
+        MM[Inc[1], pos[Inc]] = np.random.randint(delta, np.min(inc[Inc])) if len(Inc[1]) !=0 else delta 
```

Comment:
I don't fully get these two lines, `Inc[1]` refers to the columns where `inc > dec`, how come then it indexes the rows of `MM`?

Reply:
umm, I don't know if it is the best way to do it but the idea is: 

- With “Inc” and “Dec”, it detects and selects increase or decrease operator (mutation with delta)
- Then, if it's possible, when an "Inc" is higher than "Dec" It creates a matrix with deltas for increasing in mutation
- Similar with "Dec"

(*) np.where function returns a tuple, it is because of Inc[1]


@apastors Maybe I don't see any error, can you check it better?

Reply:
pos are the random positions to mutate.
In these 2 lines, it computes the difference between the maximum value and each positions. With them, it has a knowledge about how much (in terms of increments and decrements) increment or decrement

```` python
inc = (M - larvae[range(nlarvaes), pos]) 
dec = (larvae[range(nlarvaes), pos] -m) 
````

Now, it knows how much increase and decrease... but in which larvae increase and when decrease?
So, it computes when it has more margin to increase (when inc is higher than dec) or opposite

``` python
 Inc = np.where(inc>dec)  
Dec = np.where(inc<=dec) 
````

Then, inc and dec determine the limit to increase or decrease because increasing or decreasing each value it obtained the maximum or minimum M, m.
So, in order to increase or decrease without exceeding limits, a random number between a range is computed


```` python        
MM[Inc[1], pos[Inc]] = np.random.randint(delta, np.min(inc[Inc])) if len(Inc[1]) !=0 else delta 
MM[Dec[1], pos[Dec]] = -np.random.randint(delta, np.min(dec[Dec])) if len(Dec[1]) !=0 else -delta 
````

@apastors  any ideas?